util/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/AttributesUtil.java

```diff
@@ -0,0 +1,219 @@
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.HashSet;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.dcm4che3.data.Attributes;
+import org.dcm4che3.data.Sequence;
+import org.dcm4che3.data.Tag;
+import org.dcm4che3.data.VR;
+import org.dcm4che3.net.Status;
+import org.dcm4che3.net.service.DicomServiceException;
+import org.dcm4che3.util.TagUtils;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+public class AttributesUtil {
+
+  public static String getTagValue(JSONObject json, String tag) throws JSONException {
+    JSONObject jsonTag = json.getJSONObject(tag);
+    JSONArray valueArray = jsonTag.getJSONArray("Value");
+    if (valueArray.length() != 1) {
+      throw new JSONException(
+          "Expected single value for tag: "
+              + tag
+              + " in JSON:\n"
+              + valueArray.toString());
+    }
+    return valueArray.getString(0);
+  }
+
+  public static String getTagValueOrNull(JSONObject json, String tag) {
+    if (!json.has(tag)) {
+      return null;
+    }
+    JSONObject jsonTag = json.getJSONObject(tag);
+    if (!jsonTag.has("Value")) {
+      return null;
+    }
+    JSONArray valueArray = jsonTag.getJSONArray("Value");
+    if (valueArray.length() != 1) {
+      return null;
+    }
+    return valueArray.getString(0);
+  }
+
+  /**
+   * Returns array of QIDO-RS paths (more than 1 if multiple ModalitiesInStudy are present)
+   * @param attrs dcm4che Attributes to convert
+   * @param includeFields additonal includeFields for QIDO-RS
+   */
+  public static String[] attributesToQidoPathArray(Attributes attrs, String... includeFields)
+      throws IOException {
+    if (!attrs.containsValue(Tag.ModalitiesInStudy)
+        || attrs.getStrings(Tag.ModalitiesInStudy).length <= 1) {
+      return new String[]{attributesToQidoPath(attrs, includeFields)};
+    }
+
+    List<String> result = new ArrayList<>();
+    for (String modality : attrs.getStrings(Tag.ModalitiesInStudy)) {
+      Attributes attrsCopy = new Attributes(attrs);
+      attrsCopy.setString(Tag.ModalitiesInStudy, VR.CS, modality);
+      result.add(attributesToQidoPath(attrsCopy, includeFields));
+    }
+    return result.toArray(new String[]{});
+  }
+
+  /**
+   * Returns corresponding QIDO-RS path
+   * @param attrs dcm4che Attributes to convert
+   * @param includeFields additonal includeFields for QIDO-RS
+   */
+  public static String attributesToQidoPath(Attributes attrs, String... includeFields)
+      throws DicomServiceException {
+    HashSet<Integer> nonEmptyKeys = new HashSet<>();
+    for (int tag : attrs.tags()) {
+      if (attrs.containsValue(tag)) {
+        nonEmptyKeys.add(tag);
+      }
+    }
+
+    StringBuilder qidoPath = new StringBuilder();
+    if (nonEmptyKeys.contains(Tag.QueryRetrieveLevel)) {
+      switch (attrs.getString(Tag.QueryRetrieveLevel)) {
+        case "PATIENT":
```

Comment:
Patient level retrieval isn't supported by the CHC api (or QIDO I think), let's delete this case.

Reply:
Ok